### PR TITLE
Added getIOThreadPoolExecutor in HTTPServer to expose IOThreadPoolExecutor to the application

### PR DESCRIPTION
**Pull request by badarihp**
Hi,
Added getIOThreadPoolExecutor in HTTPServer so that application get access to IOThreadPoolExecutor, which helps in added observers, fetching thread pool stats etc:

Incorporated review comments from : https://github.com/facebook/proxygen/pull/228

regards
 badari

### CODE REVIEW

**Comment by badarihp** [comment]
Hi ,
     Intent is to expose IOThreadPoolExecutor to application so that
1) Application can have access to stats of thread pools
2) Application can add its own observer and 
3) Increase or Decrease  num of processing threads in IOThreadPoolExecutor at runtime.

POC git commit id : 

commit 8b7bfbd66aef9f47cf8d9ae3937070fb4bac7fc9
Author: Badari <badari.hp@gmail.com>
Date:   Thu May 17 19:01:32 2018 +0530

    For testing only, to test if num thread can increase and decrease POC


This need not be merged to main line proxygen. I could not add test code so modified EchoServer as required.

Thanks
 badari

File: proxygen/httpserver/samples/echo/EchoServer.cpp [hunk 1]
```diff
@@ -44,18 +50,45 @@ class EchoHandlerFactory : public RequestHandlerFactory {
   }
 
   RequestHandler* onRequest(RequestHandler*, HTTPMessage*) noexcept override {
+    numThreads++;
+    iotpe_->setNumThreads(numThreads.load());
```
Comment (afrind) [inline proxygen/httpserver/samples/echo/EchoServer.cpp:54]: This seems quite strange -- to increase the number of available threads on every request, with no bound?

File: proxygen/httpserver/samples/echo/EchoServer.cpp [hunk 2]
```diff
@@ -30,9 +33,12 @@ DEFINE_int32(http_port, 11000, "Port to listen on with HTTP protocol");
 DEFINE_int32(spdy_port, 11001, "Port to listen on with SPDY protocol");
 DEFINE_int32(h2_port, 11002, "Port to listen on with HTTP/2 protocol");
 DEFINE_string(ip, "localhost", "IP/Hostname to bind to");
-DEFINE_int32(threads, 0, "Number of threads to listen on. Numbers <= 0 "
+DEFINE_int32(threads, 1, "Number of threads to listen on. Numbers <= 0 "
              "will use the number of cores on this machine.");
 
+std::atomic<int> numThreads;
```
Comment (afrind) [inline proxygen/httpserver/samples/echo/EchoServer.cpp:39]: Making these global is not really ideal - would prefer they be part of something else.

File: proxygen/httpserver/samples/echo/EchoServer.cpp [hunk 3]
```diff
@@ -44,18 +50,45 @@ class EchoHandlerFactory : public RequestHandlerFactory {
...
   }
 
  private:
   folly::ThreadLocalPtr<EchoStats> stats_;
 };
 
+
+class IOCallbacks : public ThreadPoolExecutor::Observer {
+ public:
+  explicit IOCallbacks(std::shared_ptr<HTTPServerOptions> options) : options_(options) {}
```
Comment (afrind) [inline proxygen/httpserver/samples/echo/EchoServer.cpp:65]: options isn't used in this class.

**Review by afrind — CHANGES_REQUESTED** [review]
Welp this was in my review queue for 4.5 years.  I'm really sorry about that.  The core changes to HTTPServer seem ok but the echo sample changes are strange.  Since I expect this isn't urgent I'm going to close the pull for now.  But we can reopen it if you want to drive it forward.

**Comment by badarihp** [comment]
Hey Thanks Alan for response. Ill check out these comments and revert back.

**Pull request closed without merging**